### PR TITLE
Split ranges in `MergeTreeReadPoolParallelReplicasInOrder` to consume less memory during reading

### LINKED ISSUE #72022: Implement split range for parallel replicas on initiator
In `ReadFromMergeTree::spreadMarkRangesAmongStreamsWithOrder`, we split ranges to avoid reading much data. But for parallel replicas, we don't do it and comment: 
```
/// For parallel replicas the split will be performed on the initiator side.
```
https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/QueryPlan/ReadFromMergeTree.cpp#L1063

But I find it isn't implemented on the initiator side, and when we read data in reverse order in a large dataset, it will store the whole range in memory and will use a lot of memory; sometimes it will cause OOM.

So maybe we should implement split range for parallel replicas on the initiator side too.

### PR DESCRIPTION
**Pull request by JiaQiTang98**
### Changelog category (leave one):
- Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Split ranges in `MergeTreeReadPoolParallelReplicasInOrder` to consume less memory during reading.

Closes: https://github.com/ClickHouse/ClickHouse/issues/72022

### CODE REVIEW

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit bf4445158c93a6794d718f46f6102d20d125dfdb with description of existing statuses. It's updated for the latest CI running*

[✅ Click here](https://s3.amazonaws.com/clickhouse-test-reports/72173/bf4445158c93a6794d718f46f6102d20d125dfdb/ci_running.html) to open a full report in a separate page

<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs check</td><td>Builds and tests the documentation</td><td><a href="https://s3.amazonaws.com/clickhouse-test-reports/0/3ae055c74e7ce24100d3fa2348258408af8a5d1d/docs_check.html">✅ success</a></td></tr>
<tbody>
</table>
</details>

File: src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp [hunk 1]
```diff
@@ -65,21 +71,77 @@ MergeTreeReadTaskPtr MergeTreeReadPoolParallelReplicasInOrder::getTask(size_t ta
...
         {
             if (desc.info == part_info && !desc.ranges.empty())
             {
-                auto result = std::move(desc.ranges);
-                desc.ranges = MarkRanges{};
-                ProfileEvents::increment(ProfileEvents::ParallelReplicasReadMarks, desc.ranges.getNumberOfMarks());
-                return result;
+                if (mode == CoordinationMode::WithOrder)
+                {
+                    /// if already splited, just return desc.ranges
```
Comment (nickitat) [inline src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp:90]: i think it'd be simpler if we'd maintain `marks_in_range` for each part separately and calculate `split` on the fly. once `marks_in_range[part] >= max_marks_in_range` - we stop subdividing ranges.

File: src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp [hunk 2]
```diff
@@ -65,21 +71,77 @@ MergeTreeReadTaskPtr MergeTreeReadPoolParallelReplicasInOrder::getTask(size_t ta
...
+                        result.emplace_back(range.begin, range.end);
+                    }
+                    chassert(!result.empty());
+                    desc.ranges = MarkRanges{};
+                    ProfileEvents::increment(ProfileEvents::ParallelReplicasReadMarks, result.getNumberOfMarks());
+                    return result;
+                }
+                else
+                {
+                    /// for reverse order just return one range
```
Comment (nickitat) [inline src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp:120]: i don't understand why

File: src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp [hunk 3]
```diff
@@ -65,21 +71,77 @@ MergeTreeReadTaskPtr MergeTreeReadPoolParallelReplicasInOrder::getTask(size_t ta
...
-        throw Exception(ErrorCodes::LOGICAL_ERROR,
-            "Requested task with idx {}, but there are only {} parts",
-            task_idx, per_part_infos.size());
+        throw Exception(
+            ErrorCodes::LOGICAL_ERROR, "Requested task with idx {}, but there are only {} parts", task_idx, per_part_infos.size());
 
     const auto & part_info = per_part_infos[task_idx]->data_part->info;
-    auto get_from_buffer = [&]() -> std::optional<MarkRanges>
+    const auto & data_settings = per_part_infos[task_idx]->data_part->storage.getSettings();
+    auto get_from_buffer = [&,
```
Comment (nickitat) [inline src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp:79]: i'd also check that we don't exceed `min_marks_per_task`

Comment (nickitat) [inline src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp:79]: let's create a test for mem usage that reproduces the problem

File: src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp (hunk 1 above)
Comment (JiaQiTang98) [reply src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp:90]: Great idea, i will change it.

File: src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp (hunk 2 above)
Comment (JiaQiTang98) [reply src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp:120]: In `MergeTreeInReverseOrderSelectAlgorithm::readFromTask`, clickhouse will read all data in `MergeTreeReadTask`. 

https://github.com/ClickHouse/ClickHouse/blob/master/src/Storages/MergeTree/MergeTreeSelectAlgorithms.cpp#L45

So I think if we still return all MarkRanges to create a task, `MergeTreeInReverseOrderSelectAlgorithm` will still store the whole ranges in memory. So i just return one range to create a task

File: src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp (hunk 3 above)
Comment (JiaQiTang98) [reply src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp:79]: I think `min_marks_per_task` has already been processed in InOrderCoordinator::handleRequest and maybe we don't need to check it here.

Comment (JiaQiTang98) [reply src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp:79]: I will add some tests

File: src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp (hunk 2 above)
Comment (nickitat) [reply src/Storages/MergeTree/MergeTreeReadPoolParallelReplicasInOrder.cpp:120]: maybe it is ok if we do it once for each part (task with 1 mark, then with 2, then with 4 until we reach `min_marks_per_task`), but then we shouldn't reset `marks_in_range` back to 1

**Comment by JiaQiTang98** [comment]
@nickitat Hi, Could you help to review it again?

File: tests/queries/0_stateless/03272_parallel_replicas_read_in_order.sql [hunk 4]
```diff
@@ -0,0 +1,21 @@
+SET log_queries = 1;
+SET optimize_read_in_order=1;
+DROP TABLE IF EXISTS read_in_order_with_parallel_replicas;
+CREATE TABLE read_in_order_with_parallel_replicas(id UInt64) ENGINE=MergeTree ORDER BY id SETTINGS index_granularity=1;
+
+
+INSERT INTO read_in_order_with_parallel_replicas SELECT number from system.numbers limit 100000;
+
+SELECT * from read_in_order_with_parallel_replicas ORDER BY id desc limit 1 settings max_threads=1;
```
Comment (nickitat) [inline tests/queries/0_stateless/03272_parallel_replicas_read_in_order.sql:9]: let's also test read in asc order

**Comment by nickitat** [comment]
apart from test - lgtm

**Review by nickitat — APPROVED** [review]

**Pull request merged**